Cloud-native architecture (#4)
SnellerInc/blogs: cloud-native-architecture.md

# Sneller cloud-native architecture

> Most database engines were designed well before cloud-native architecture become popular. Some engines have been migrated to run in the cloud, but the architecture is often still monolithic.

Reviewer: cloud-native architecture --> cloud-native architectures

> In the beginning, databases could only scale vertically (adding more CPU power). Vertical scaling is limited, so databases started to scale horizontally (adding more computers). Although the database can scale horizontally, each node typically has the same responsibilities:

Reviewer: consider: adding more computers --> adding more servers

> In the beginning, databases could only scale vertically (adding more CPU power). Vertical scaling is limited, so databases started to scale horizontally (addming more computers). Although the database can scale horizontally, all nodes have the same responsibilities. Each node typically takes care of:

Reviewer 2: ```suggestion
In the beginning, databases could only scale vertically (adding more CPU power). Vertical scaling is limited, so databases started to scale horizontally (adding more computers). Although the database can scale horizontally, each node typically has the same responsibilities:
```

1. Ingesting the data.
2. Storing the data.
3. Excute queries on the data.

Each nodes performs all these tasks:
```mermaid
flowchart LR
    A[Ingestion<br>Storage<br>Query Execution] --- B[Ingestion<br>Storage<br>Query Execution]
    B --- C[Ingestion<br>Storage<br>Query Execution]
```

> Direct-attached storage is often used to persist the data. If the node is removed from the cluster, so is the data on that node. As clusters grow over time, they becomes increasingly difficult to manage. Operational costs skyrocket and dedicated personnel are required to manage the clusters.

Reviewer: they becomes --> they become

> Direct-attached storage is often used to persist the data. If the node is removed from the cluster, so is the data on that node. As clusters grow over time, they becomes increasingly difficult to manage. Operational costs skyrocket and dedicated personnel are required to manage the clusters.

Reviewer 2: > If the node is removed from the cluster, so is the data on that node.

This is probably too much of a simplification.
For systems like Elastic, CockroachDB, Casandra, etc. that are designed to run multi-node, there is typically a re-balancing algorithm for moving data to the right place. For single-master DBs like Postgres you still have read replicas, and even if you spin down the primary instance you still have an EBS snapshot of the disk unless you told EC2 to delete the snapshot on termination.

> Direct-attached storage is often used to persist the data. If the node is removed from the cluster, so is the data on that node. When clusters grow bigger over time, this becomes increasingly more difficult to manage. Operational costs skyrocket and dedicated personel is required to manage the clusters.

Reviewer 2: ```suggestion
Direct-attached storage is often used to persist the data. If the node is removed from the cluster, so is the data on that node. As clusters grow over time, they becomes increasingly difficult to manage. Operational costs skyrocket and dedicated personnel are required to manage the clusters.
```

> Sneller uses a different approach that fits cloud-native architecture much better. It clearly seperates the various aspects of the database.

Reviewer: cloud-native architecture --> cloud-native architectures

Reviewer 2: "cloud-native architecture" is a proper noun, e.g. ["Principles for Cloud-Native Architecture](https://cloud.google.com/blog/products/application-development/5-principles-for-cloud-native-architecture-what-it-is-and-how-to-master-it), so it shouldn't be pluralized

Reviewer: Ah, didn't know that.

```mermaid
flowchart TD
    subgraph I[Ingestion]
        direction LR
        A[Node]
    end
    subgraph QE[Query execution]
        direction LR
        X[Node 1] --- Y[Node 2] --- Z[Node 3]
    end
    I --> OS(((Object Storage))) --> QE
```

> Each component runs autonomous and can be scaled independently.

Reviewer 2: ```suggestion
Each component runs autonomously and can be scaled independently.
```

1. Ingestion (loading data into the cluster) can run on any node.
2. Persistent data is stored in object storage (i.e. S3 or Minio).
3. Query execution is performed by specialized nodes that typically have a lot of RAM.

> This seperation has some huge advantages that can reduce operational cost and maintenance significantly.

Reviewer 2: ```suggestion
This separation has some huge advantages that can reduce operational and maintenance costs significantly.
```

## Ingestion

> Raw JSON is not efficient to store and process. That's why Sneller converts all ingested data to a more efficient open format. This format is using open-source technologies ([binary ION](https://amzn.github.io/ion-docs/docs/binary.html) and [Zstandard compression](https://facebook.github.io/zstd/)), so there is no vendor lock-in.

Reviewer 2: I'd rather avoid the parenthetical remarks if we can avoid them. Either the clarifications deserve their own independent clauses or they don't add enough information to be worth the additional space.

> Raw JSON is not efficient to store and process. That's why Sneller converts all ingested data to a more efficient open format. This format is using open-source technologies ([binary ION](https://amzn.github.io/ion-docs/docs/binary.html) and [Zstandard compression](https://facebook.github.io/zstd/)), so there is no vendor lock-in.

Author: I added *Conceptually, it is a compressed row-based binary JSON format.* to the paragraph.

```mermaid
flowchart LR
    JSON[JSON] --> Parser[/JSON parser/] 
    Parser --> Converter[/ION conversion/]
    Converter --> Compress[/compress/]
    Compress --> Persist((object storage))
```

> Ingested data is smaller and can easily converted back to the original JSON. There is no need to keep the original JSON files saving a lot of storage costs. Compressed binary ION is typically 5-30x more storage efficient compared to raw JSON files.

Reviewer 2: ```suggestion
Ingested data is smaller and can easily converted back to the original JSON. There is no need to keep the original JSON files, so storage costs are lower. Compressed binary ION is typically 5-30x more space-efficient than raw JSON files.
```

Sneller is built for performance, so we built a custom-built JSON parser and binary ION writer. Ingestion is fast and efficient, so it can typically run on relatively low-end hardware and has low memory requirements. We can ingest up to **TODO** MiB/sec on a single **TODO** CPU.

> The ingestion also supports ingesting other formats, such as CSV. It can easily be extended to support other formats.

Reviewer 2: Not yet, though.

Author: Oh, I thought CSV was always supported. I'll change it.

> When the data-set is static, then the ingestion node can be removed after all the data has been ingested.

Reviewer 2: ```suggestion
If the data-set is static, the ingestion node can be shut down after all the data has been ingested.
```

## Persisting to object storage

> All persistent data in Sneller is stored in object storage (i.e. S3 or Minio). This has quite some advantages:

Reviewer 2: ```suggestion
All persistent data in Sneller is stored in object storage (i.e. S3 or Minio). This has a number of advantages:
```

1. You get all the benefits of the underlying object storage. AWS S3 provides high availability, security and up to 99.999999999% durability out of the box.

> 2. No need to scale your database cluster and add expensive nodes, because you need additional storage.

Reviewer 2: ```suggestion
2. No need to scale your database cluster and add expensive nodes when you need additional storage.
```

> 3. No loss of data, when the database cluster is deleted.

Reviewer 2: ```suggestion
3. No loss of data when the database cluster is deleted.
```

> 4. No need to learn new storage. Just use the object storage you are already familiar with.

Reviewer 2: ```suggestion
4. No need to learn a new storage technology. Just use the object storage you are already familiar with.
```

> Moving persistent data from the database engine to object storage significantly reduces operational costs and risks. To ensure maximum performance, the object storage should be located as close to the query engine as possible (i.e. same AWS region).

Reviewer: as close to --> near

> Moving persistent data from the database engine to object storage significantly reduce operational costs and risks. To make sure performance is high, the object storage should be located as close to the query engine as possible (i.e. same AWS region).

Reviewer 2: ```suggestion
Moving persistent data from the database engine to object storage significantly reduces operational costs and risks. To ensure maximum performance, the object storage should be located as close to the query engine as possible (i.e. same AWS region).
```

> Although Sneller currently only supports AWS S3 (or compatible), the access-layer is abstracted and can easily be ported to support Azure Blob storage, Google Cloud Storage, ...

Reviewer 2: ```suggestion
Although Sneller currently only supports AWS S3 (or compatible) object storage, the access-layer is abstracted and can easily be ported to support Azure Blob storage, Google Cloud Storage, ...
```

## Query execution

> Sneller's query execution engine uses AVX-512 to run queries with much higher throughput compared to traditional query engines. Queries can fan-out over multiple nodes with lineair scalability. Data is retrieved from object storage and is stored in a transient compressed memory cache on the node to reduce network bandwidth.

Reviewer 2: ```suggestion
Sneller's query execution engine uses AVX-512 to run queries with much higher throughput compared to traditional query engines. Queries can fan-out over multiple nodes with lineair scalability. Data is retrieved from object storage and is stored in a transient compressed memory cache on the node to reduce network utilization under load.
```

> AWS provides up to 100Gb/s networking speed and data is fetched in parallel from S3 to fully utilize the available network bandwidth. Even if data is not in cache, then it can be loaded fast and efficient.

Reviewer 2: ```suggestion
AWS provides up to 100Gb/s of bandwidth per EC2 instance and data is fetched in parallel from S3 to fully utilize the available network bandwidth. Even if data is not in cache, it can be loaded quickly and efficiently.
```

> You can scale down (or complete terminate) the query execution nodes, when you don't need to query data.No persistent data is stored in the query nodes, so there is no loss of data. Ingestion of new data can still continue, because it runs on seperate nodes. This can save even more operational costs, because database nodes are often expensive, due to large amount of RAM.

Reviewer 2: ```suggestion
You can scale down (or completely terminate) the query execution nodes when you don't need to query data. No persistent data is stored in the query nodes, so there is no loss of data. Ingestion of new data can still continue, because it runs on separate nodes. This can save even more operational costs, because database nodes are often expensive, due to large amount of RAM.
```
